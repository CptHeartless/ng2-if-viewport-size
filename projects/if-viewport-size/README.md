# IfViewportSize

This library was generated with [Angular CLI](https://github.com/angular/angular-cli) version 8.2.11.

## Code scaffolding

Run `ng generate component component-name --project if-viewport-size` to generate a new component. You can also use `ng generate directive|pipe|service|class|guard|interface|enum|module --project if-viewport-size`.
> Note: Don't forget to add `--project if-viewport-size` or else it will be added to the default project in your `angular.json` file. 

## Build

Run `ng build if-viewport-size` to build the project. The build artifacts will be stored in the `dist/` directory.

## Publishing

After building your library with `ng build if-viewport-size`, go to the dist folder `cd dist/if-viewport-size` and run `npm publish`.

## Running unit tests

Run `ng test if-viewport-size` to execute the unit tests via [Karma](https://karma-runner.github.io).

## Further help

To get more help on the Angular CLI use `ng help` or go check out the [Angular CLI README](https://github.com/angular/angular-cli/blob/master/README.md).
