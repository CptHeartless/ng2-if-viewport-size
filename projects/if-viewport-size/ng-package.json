{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../dist/if-viewport-size",
  "lib": {
    "entryFile": "src/public-api.ts"
  }
}