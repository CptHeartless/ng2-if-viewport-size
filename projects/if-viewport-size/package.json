{
  "name": "ng2-if-viewport-size",
  "version": "1.0.0",
  "description": "Angular2+ directive that able to hide/show component according to the viewport size.",
  "main": "./bundles/if-viewport-size.umd.js",
  "peerDependencies": {
    "@angular/common": "^8.2.11",
    "@angular/core": "^8.2.11",
    "rxjs": "~6.4.0"
  }
}
