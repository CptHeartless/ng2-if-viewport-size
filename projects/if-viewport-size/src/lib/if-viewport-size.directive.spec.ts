import {
  Component,
  ElementRef,
  Injectable,
  TemplateRef,
  ViewContainerRef,
} from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { Observable, Subscription } from 'rxjs';
import { IfViewportSizeDirective } from './if-viewport-size.directive';
import { IfViewportSizeService, State } from './if-viewport-size.service';

@Component({
  template: `
    <span *ifViewportSize="'small'">Small viewport</span>
    <span *ifViewportSize="'medium'">Medium viewport</span>
    <span *ifViewportSize="'large'">Large viewport</span>
    <span *ifViewportSize="['small', 'medium']">Small and medium viewport</span>
    <span *ifViewportSize="['medium', 'large']">Large and medium viewports</span>
  `,
})
class TestComponent { }

@Injectable()
class IfViewportSizeServiceStub {
  state: State = 'small';
  state$: Observable<State> = new Observable(() => {});
}

describe('[IfViewportSizeDirective] DOM', () => {
  let fixture: ComponentFixture<TestComponent>;
  let hostElement: ElementRef;
  let service: IfViewportSizeServiceStub;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [TestComponent, IfViewportSizeDirective],
      providers: [
        {
          provide: IfViewportSizeService,
          useValue: new IfViewportSizeServiceStub(),
        },
      ],
    });

    fixture = TestBed.createComponent(TestComponent);
    hostElement = fixture.nativeElement;
    service = TestBed.get(IfViewportSizeService) as IfViewportSizeServiceStub;
  });

  test.each([
    ['Small Viewport', 'small'],
    ['Medium Viewport', 'medium'],
    ['Large Viewport', 'large'],
  ])('rendering for %s', (_, state: State) => {
    service.state = state;
    fixture.detectChanges();

    expect(hostElement)
      .toMatchSnapshot();
  });
});

describe('IfViewportSizeDirective {}', () => {
  // Directive dependencies
  let viewContainerStub: Partial<ViewContainerRef>;
  let templateRefStub: Partial<TemplateRef<any>>;
  let serviceFixture: IfViewportSizeServiceStub;

  // Directive subscription teardown spies
  let onSubSpy: jest.SpyInstance<Subscription, Array<(state: State) => void>>;
  let onUnsubSpy: jest.SpyInstance<void, Array<() => void>>;

  // Directive instance
  let instance: IfViewportSizeDirective;

  beforeEach(() => {
    serviceFixture = new IfViewportSizeServiceStub();

    const subFixture = new Subscription();
    onSubSpy = jest.spyOn(serviceFixture.state$, 'subscribe')
      .mockReturnValue(subFixture);
    onUnsubSpy = jest.spyOn(subFixture, 'unsubscribe');

    viewContainerStub = {
      createEmbeddedView: jest.fn(),
      clear: jest.fn(),
    };
    templateRefStub = {};

    instance = new IfViewportSizeDirective(
      serviceFixture as IfViewportSizeService,
      templateRefStub as TemplateRef<ElementRef>,
      viewContainerStub as ViewContainerRef,
    );

    instance.ifViewportSize = 'small';
  });

  test('.ngOnInit() when not visible', () => {
    serviceFixture.state = 'medium';

    instance.ngOnInit();

    expect(viewContainerStub.clear)
      .toBeCalled();
    expect(serviceFixture.state$.subscribe)
      .toBeCalled();
  });

  test('.ngOnInit() when visible', () => {
    serviceFixture.state = 'small';

    instance.ngOnInit();

    expect(viewContainerStub.createEmbeddedView)
      .toBeCalledWith(templateRefStub);
    expect(onSubSpy)
      .toBeCalled();
  });

  test('.ngOnDestroy()', () => {
    instance.ngOnInit();

    expect(onUnsubSpy).not
      .toBeCalled();
    instance.ngOnDestroy();
    expect(onUnsubSpy)
      .toBeCalled();
  });

  it('should be hidden on event', () => {
    instance.ngOnInit();
    const handler = onSubSpy.mock.calls[0][0];

    expect(viewContainerStub.clear)
      .toHaveBeenCalledTimes(0);
    handler('large');
    expect(viewContainerStub.clear)
      .toHaveBeenCalledTimes(1);
  });

  test('should be shown on event', () => {
    serviceFixture.state = 'large';
    instance.ngOnInit();
    const handler = onSubSpy.mock.calls[0][0];

    expect(viewContainerStub.createEmbeddedView)
      .toHaveBeenCalledTimes(0);
    handler('small');
    expect(viewContainerStub.createEmbeddedView)
      .toHaveBeenCalledTimes(1);
    expect(viewContainerStub.createEmbeddedView)
      .toBeCalledWith(templateRefStub);
  });
});
