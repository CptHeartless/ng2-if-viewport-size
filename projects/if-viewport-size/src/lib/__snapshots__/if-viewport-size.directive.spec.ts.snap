// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[IfViewportSizeDirective] DOM rendering for Large Viewport 1`] = `
<div
  id="root2"
  ng-version="8.2.11"
>
  
  
  
  <span>
    Large viewport
  </span>
  
  
  <span>
    Large and medium viewports
  </span>
</div>
`;

exports[`[IfViewportSizeDirective] DOM rendering for Medium Viewport 1`] = `
<div
  id="root1"
  ng-version="8.2.11"
>
  
  
  <span>
    Medium viewport
  </span>
  
  
  <span>
    Small and medium viewport
  </span>
  
  <span>
    Large and medium viewports
  </span>
</div>
`;

exports[`[IfViewportSizeDirective] DOM rendering for Small Viewport 1`] = `
<div
  id="root0"
  ng-version="8.2.11"
>
  
  <span>
    Small viewport
  </span>
  
  
  
  <span>
    Small and medium viewport
  </span>
  
</div>
`;
